oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/IndexUtils.java

```diff
@@ -241,4 +242,37 @@ public static String getAsyncLaneName(NodeState idxState, String indexPath, Prop
         }
         return null;
     }
+
+    /**
+     * Retrieves the calling class and method from the call stack; this is determined by unwinding
+     * the stack until it finds a combination of full qualified classname + method (separated by ".") which
+     * do not start with any of the values provided by the ignoredJavaPackages parameters. If the provided
+     * parameters cover all stack frames, the whole query is considered to be internal, where the 
+     * actual caller doesn't matter (or cannot be determined clearly). In this case a short message
+     * indicating this is returned.
+     *
+     * If the ignoredJavaPackages parameter is null or empty, the caller is not looked up, but
+     * instead it is assumed, that the feature is not configured; in that case a short messages
+     * is returned indicating that the feature is not configured.
+     *
+     * @param ignoredJavaPackages the java packages or class names
+     * @return the calling class or another non-null value
+     */
+    @NotNull
+    public static String getCaller(@Nullable String[] ignoredJavaPackages) {
+        if (ignoredJavaPackages == null || ignoredJavaPackages.length == 0) {
+            return "(<function not configured>)";
+        }
+
+        // With java9 we would use https://docs.oracle.com/javase/9/docs/api/java/lang/StackWalker.html
+        final StackTraceElement[] callStack = Thread.currentThread().getStackTrace();
+        for (StackTraceElement stackFrame : callStack) {
```

Comment:
This means we are capturing the top of the stack first. Wondering if we would be interested in bottom of the stack or top.

Reply:
We are interested in the first class name, which is is not ignored. This should give us a good understanding which code is triggering a query (assuming that we put all relevant packages/classes into the ```ignoredJavaPackages``` which do not help us to determine that).

Reply:
Yes, my question was more around - are we trying to obtain the generic feature which caused this query or, the api which actually executed this query. The stack would be something like -
```
---other ignored packages---
ignored oak query classes
some_api_used_by_feature_classes_to_execute_query
high_level_feature_classes
---ignored packages---
```
Iterating the stack from top to bottom and capturing the first non-ignored class would provide the api used by feature classes, not exactly those features which are using this query.

Reply:
Of course it is possible to also put ```some_api_used_by_feature_classes_to_execute_query``` (packages and/or classes) to the ```ignoredJavaPackages```. That probably requires some fine-tuning, but it is definitely possible.

When testing with AEM I used this list:
```
org.apache.jackrabbit.oak
java.lang
org.apache.sling.jcr.resource.internal.helper.JcrResourceUtil.query
org.apache.sling.jcr.resource.internal.helper.jcr.BasicQueryLanguageProvider
org.apache.sling.resourceresolver.impl.providers.stateful.AuthenticatedResourceProvider.findResources
org.apache.sling.resourceresolver.impl.helper.ResourceResolverControl.findResources
org.apache.sling.resourceresolver.impl.ResourceResolverImpl.findResources
com.day.cq.search.impl.builder.QueryImpl
```

and got good results.

Reply:
okay, if the results are good, then let's go ahead :)